### PR TITLE
Feature/5208

### LINKED ISSUE #5208: Add custom buttons in Buttons Toolbar
**Description**  
I would like to add some custom buttons on the right on the bootstrap-table ones. I see you can add custom html [below the buttons](https://examples.bootstrap-table.com/index.html#options/buttons-toolbar.html), I'm asking if it is possible to add buttons on the same line. 

**Example**  
Please see screenshot
Bootstrap-Table-Examples

### PR DESCRIPTION
**Pull request by UtechtDustin**
**Description**
This PR will added the ability to add custom Buttons to the "buttonbar".
The button is configurable:
- Buttontext
- Buttonicon
- Event
- Enabled
- Htmlattributes

It will also allow to sort ALL button (including the buttons of extensions and the custom Buttons)

**Todos**

- [x] Add the ability to add custom buttons
- [x] Add the buttons of the bootstrap-table core to the new system
- [x] Add the buttons of the bootstrap-table extensions to the new system
- [x] Add documentation
- [x] Add examples 

**Bug fix?**
no

**New Feature?**
yes

**Resolve an issue?**
Fix #5208

**Example(s)?**
Custom Button: https://live.bootstrap-table.com/code/UtechtDustin/4184
Custom Button with text: https://live.bootstrap-table.com/code/UtechtDustin/4185
Custom Button only text: https://live.bootstrap-table.com/code/UtechtDustin/4189
Custom Button with event: https://live.bootstrap-table.com/code/UtechtDustin/4188
Custom Button with multiple events (and custom eventtypes): https://live.bootstrap-table.com/code/UtechtDustin/4209
Custom Button which is ordered before the `toggle columns` Button: https://live.bootstrap-table.com/code/UtechtDustin/4194
Custom Button with html attributes: https://live.bootstrap-table.com/code/UtechtDustin/4191
Custom Button which is added but disabled: https://live.bootstrap-table.com/code/UtechtDustin/4192
Custom button which is added (disabled) but can be enabled with `data-show-<buttonName>`: https://live.bootstrap-table.com/code/UtechtDustin/4193 
Custom button with custom html: https://live.bootstrap-table.com/code/UtechtDustin/4207
Multiple custom buttons: https://live.bootstrap-table.com/code/UtechtDustin/4196
Full Example: https://live.bootstrap-table.com/code/UtechtDustin/4195

### CODE REVIEW

**Comment by antonioaltamura** [comment]
@UtechtDustin I've taken a look to your examples (not yet at the implementation). It looks nice, but I've got 2 notes:
- I think you should stick with the Bootstrap table [Column Event](https://examples.bootstrap-table.com/index.html#column-options/events.html) interface. E.g. `'click button': function (e, value, row, index) {...}`
In your PR, what if I want to handle the `mouseover` event? https://live.bootstrap-table.com/code/UtechtDustin/4188 
- I think it would be nice to allow the user, in case he needs it, to add total custom html code, not just a button through some options.

**Comment by UtechtDustin** [comment]
@antonioaltamura 
> I think you should stick with the Bootstrap table Column Event interface. E.g. 'click button': function (e, value, row, index) {...}
In your PR, what if I want to handle the mouseover event? https://live.bootstrap-table.com/code/UtechtDustin/4188

Could you describe a usecase for a mouseover event ?
If that makes sense i will add a option to set the event.

> I think it would be nice to allow the user, in case he needs it, to add total custom html code, not just a button through some options.

Its already possible, i forgott to create a example: https://live.bootstrap-table.com/code/UtechtDustin/4207 :-)

**Comment by UtechtDustin** [comment]
@antonioaltamura 
> I think you should stick with the Bootstrap table Column Event interface. E.g. 'click button': function (e, value, row, index) {...}
In your PR, what if I want to handle the mouseover event? https://live.bootstrap-table.com/code/UtechtDustin/4188

You now have multiple ways to define event(s):
- you can use the name of the function ([example](https://live.bootstrap-table.com/code/UtechtDustin/4210))
- you can use the function ([example](https://live.bootstrap-table.com/code/UtechtDustin/4211))
- you can use a object with the name of the function ([example](https://live.bootstrap-table.com/code/UtechtDustin/4209))
- you can use a object with functions ([example](https://live.bootstrap-table.com/code/UtechtDustin/4212))
- you can use a object mixed with name/function ([example](https://live.bootstrap-table.com/code/UtechtDustin/4213))

File: site/docs/extensions/auto-refresh.md [hunk 1]
```diff
@@ -76,4 +76,4 @@ toc: true
 
 - **Parameter:** `undefined`
 
-- **Default:** `'Auto Refresh'`
+- **Default:** `function () { return "Auto Refresh" }`
```
Comment (djhvscf) [inline site/docs/extensions/auto-refresh.md:79]: How will the localize works here?

Comment (UtechtDustin) [reply site/docs/extensions/auto-refresh.md:79]: What do you mean ?
I just added missing locales and changed other locales to that format for consistency.

**Comment by UtechtDustin** [comment]
> Why in this example: https://live.bootstrap-table.com/code/UtechtDustin/4192 the enable = false is not displaying the button? I think it should display it but disabled.

I guess we should rename it, because its not to show a disables button its to hide it by default and can be shown again via data attribute (like all our other buttons).
Do you have a good idea for that name ? `autoShow` ?

> In this example: https://live.bootstrap-table.com/code/djhvscf/4220 the data-show-btntest1="false" is not working. Am i missing something?

Currently its not implement to use the data attribute to hide a button, because we already have the `enabled` option (which must be renamed). But i don implement it to prevent confusion.

**Comment by UtechtDustin** [comment]
> Why in this example: https://live.bootstrap-table.com/code/UtechtDustin/4192 the enable = false is not displaying the button? I think it should display it but disabled.

Renamed the option from `enabled` to `render`.

> In this example: https://live.bootstrap-table.com/code/djhvscf/4220 the data-show-btntest1="false" is not working. Am i missing something?

Works now.

**Comment by djhvscf** [comment]
I will review what's happening with the compilation error

**Comment by UtechtDustin** [comment]
@wenzhixin already fixed that on a other PR once, but im not sure why this is happening again.

**Comment by UtechtDustin** [comment]
@wenzhixin could you check that ?

**Comment by wenzhixin** [comment]
OK, I will check it.

**Comment by wenzhixin** [comment]
The paginationSwitch button always display by default: https://live.bootstrap-table.com/code/wenzhixin/4402

**Comment by djhvscf** [comment]
@UtechtDustin do you need any help here?

**Comment by UtechtDustin** [comment]
@djhvscf if you have time sure.
Im currently busy and dont have that much time right now.

**Comment by djhvscf** [comment]
> The paginationSwitch button always display by default: https://live.bootstrap-table.com/code/wenzhixin/4402

The problem is that when the code is iterating throug the this.buttons properties, it is applying lowerCase so that's why the showPaginationSwich is not working. continue working on this

**Comment by djhvscf** [comment]
> > The paginationSwitch button always display by default: https://live.bootstrap-table.com/code/wenzhixin/4402
> 
> The problem is that when the code is iterating throug the this.buttons properties, it is applying lowerCase so that's why the showPaginationSwich is not working. continue working on this

This is fixed now. Working on why the compilation is failing. FYI

**Comment by djhvscf** [comment]
After several tries... it is working. @wenzhixin  @UtechtDustin  please take a look and approve it.

**Comment by UtechtDustin** [comment]
@djhvscf it seems that the `enabled` option dont work after your patches.
Check the following examples (same as from the first comment):
Custom Button which is added but disabled: https://live.bootstrap-table.com/code/UtechtDustin/4192
Custom button which is added (disabled) but can be enabled with data-show-<buttonName>: https://live.bootstrap-table.com/code/UtechtDustin/4193

**Review by djhvscf — APPROVED** [review]

**Comment by UtechtDustin** [comment]
Oh it seems i failed the force push, but i see your fixing it :+1:

**Review by wenzhixin — APPROVED** [review]

**Pull request merged**